src/icastats_shared.c

```diff
@@ -46,74 +44,77 @@ volatile int stats_shm_handle = NOT_INITIALIZED;
 
 int stats_mmap(int user)
 {
+	int stats_shm_handle;
 	char shm_id[NAME_LENGHT];
 	struct stat stat_buf;
 
-	if (stats == NULL) {
-		sprintf(shm_id, "icastats_%d",
-			user == -1 ? geteuid() : (uid_t)user);
+	if (stats != NULL)
+		return 0;
 
-		stats_shm_handle = shm_open(shm_id, O_RDWR,  S_IRUSR | S_IWUSR);
-		if (stats_shm_handle == NOT_INITIALIZED)
-			stats_shm_handle = shm_open(shm_id, O_CREAT | O_RDWR,
-						    S_IRUSR | S_IWUSR);
+	sprintf(shm_id, "icastats_%d",
+		user == -1 ? geteuid() : (uid_t)user);
 
-		if (stats_shm_handle == NOT_INITIALIZED)
-			return -1;
+	stats_shm_handle = shm_open(shm_id,
+				    O_CREAT | O_RDWR,
+				    S_IRUSR | S_IWUSR);
 
-		if (user > 0 && geteuid() == 0) {
-			if (fchown(stats_shm_handle, user, user) == -1) {
-				close(stats_shm_handle);
-				return -1;
-			}
-		}
+	if (stats_shm_handle == -1)
+		return -1;
 
-		if (fstat(stats_shm_handle, &stat_buf)) {
-			close(stats_shm_handle);
-			return -1;
-		}
+	if ((user > 0 && geteuid() == 0) &&
+	    (fchown(stats_shm_handle, user, user) == -1))
+		goto err;
 
-		if (ftruncate(stats_shm_handle, STATS_SHM_SIZE) == -1) {
-			close(stats_shm_handle);
-			return -1;
-		}
+	if (fstat(stats_shm_handle, &stat_buf))
+		goto err;
 
-		stats = (stats_entry_t *) mmap(NULL, STATS_SHM_SIZE, PROT_READ |
-						 PROT_WRITE, MAP_SHARED,
-						 stats_shm_handle, 0);
-		if (stats == MAP_FAILED){
-			close(stats_shm_handle);
-			stats = NULL;
-			return -1;
-		}
+	if (ftruncate(stats_shm_handle, STATS_SHM_SIZE) == -1)
```

Comment:
We should not do that if we did not generate the segment.  What if - for whatever reason - two different versions of libica are running in the system with different `STATS_SHM_SIZE` values?

Reply:
This might be a valid point, but please keep in mind, that my change just removed one level of indention and the curly brackets. This is the only reason, why the line appeared in this commit.

Reply:
A fix for this potential issue should anyhow go into a separate commit.